src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -90,44 +90,92 @@ public Dictionary(int capacity, IEqualityComparer<TKey>? comparer)
-                int count = d._count;
-                Entry[]? entries = d._entries;
+                Dictionary<TKey, TValue> source = (Dictionary<TKey, TValue>)dictionary;
+
+                if (source.Count == 0)
+                {
+                    // Nothing to copy, all done
+                    return;
+                }
+
+                Debug.Assert(source._buckets is not null);
+                Debug.Assert(source._entries is not null);
+
+                Entry[] oldEntries = source._entries;
+                // If comparers are the same, we can copy dictionary without rehashing.
+                if (source._comparer == dest._comparer)
+                {
+                    Entry[] newEntries = dest._entries;
+                    int oldCount = source._count;
+                    int newCount = 0;
+                    for (int i = 0; i < oldCount; i++)
+                    {
+                        uint hashCode = oldEntries[i].hashCode;
+                        // Only copy if an entry
+                        if (oldEntries[i].next >= -1)
+                        {
+                            ref Entry entry = ref newEntries[newCount];
+                            entry = oldEntries[i];
+                            ref int bucket = ref dest.GetBucket(hashCode);
+                            entry.next = bucket - 1; // Value in _buckets is 1-based
+                            bucket = newCount + 1;
+                            newCount++;
+                        }
+                    }
+
+                    dest._count = newCount;
+                    return;
+                }
+
+                // Comparers differ need to rehash all the entires via Add
+                int count = source._count;
                 for (int i = 0; i < count; i++)
                 {
-                    if (entries![i].next >= -1)
+                    // Only copy if an entry
+                    if (oldEntries[i].next >= -1)
                     {
-                        Add(entries[i].key, entries[i].value);
+                        dest.Add(oldEntries[i].key, oldEntries[i].value);
                     }
                 }
                 return;
             }
 
-            foreach (KeyValuePair<TKey, TValue> pair in dictionary)
-            {
-                Add(pair.Key, pair.Value);
-            }
+            // Fallback path for IDictionary that isn't a non-subclassed Dictionary<TKey,TValue>.
+            CopyFallback(dictionary, dest);
         }
 
-        public Dictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection) : this(collection, null) { }
-
-        public Dictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection, IEqualityComparer<TKey>? comparer) :
-            this((collection as ICollection<KeyValuePair<TKey, TValue>>)?.Count ?? 0, comparer)
+        private static void CopyFallback(IEnumerable<KeyValuePair<TKey, TValue>> collection, Dictionary<TKey, TValue> dest)
```

Comment:
This is used from single place. I do not think it is is worth it to factor it out to separate method.

When I am looking at code size, my primary concern is AOT code size.

Reply:
Factored back in